src/urql/atomWithQuery.ts

```diff
@@ -0,0 +1,132 @@
+  opts?: Partial<OperationContext>
+}
+
+type QueryArgs<Data, Variables extends object> = {
+  query: TypedDocumentNode<Data, Variables>
+  variables?: Variables
+  requestPolicy?: RequestPolicy
+  context?: Partial<OperationContext>
+  pause?: boolean
+}
+
+export function atomWithQuery<Data, Variables extends object>(
+  createQueryArgs: (get: Getter) => QueryArgs<Data, Variables>
+): WritableAtom<[Data, GraphQLExtensions], ResultActions> {
+  const operationResultAtom = atom<
+    OperationResult<Data, Variables> | Promise<OperationResult<Data, Variables>>
+  >(
+    new Promise<OperationResult<Data, Variables>>(() => {}) // infinite pending
+  )
+  const queryAtom = atom<
+    [QueryArgs<Data, Variables>, WritableAtom<null, any>],
+    ResultActions
+  >(
+    (get) => {
+      const args = createQueryArgs(get)
+      const initAtom = atom(
+        null,
+        (
+          get,
+          set,
+          action:
+            | { type: 'init'; initializer: (client: Client) => void }
+            | { type: 'result'; result: OperationResult<Data, Variables> }
+        ) => {
+          if (action.type === 'init') {
+            set(
+              operationResultAtom,
+              new Promise<OperationResult<Data, Variables>>(() => {}) // new fetch
+            )
+            if (!args.pause) {
+              action.initializer(getClient(get, set))
+            }
+          } else if (action.type === 'result') {
+            if (!action.result.stale) {
+              set(operationResultAtom, action.result)
+            }
+          }
+        }
+      )
+      initAtom.onMount = (dispatch) => {
+        let unsub: (() => void) | undefined | false
+        const initializer = (client: Client) => {
+          const subscription = pipe(
+            client.query(args.query, args.variables, {
+              requestPolicy: args.requestPolicy,
+              ...args.context,
+            }),
+            subscribe((result) => {
+              dispatch({ type: 'result', result })
+            })
+          )
+          if (unsub === false) {
+            subscription.unsubscribe()
+          } else {
+            unsub = () => {
+              subscription.unsubscribe()
+            }
+          }
+        }
+        dispatch({ type: 'init', initializer })
+        return () => {
+          if (unsub) {
+            unsub()
+          }
+          unsub = false
+        }
+      }
+      return [args, initAtom]
+    },
+    (get, set, action) => {
+      if (action.type === 'execute') {
+        const [args] = get(queryAtom)
+        const client = getClient(get, set)
+        client
+          .query(args.query, args.variables, {
+            requestPolicy: args.requestPolicy,
+            ...args.context,
+            ...action.opts,
+          })
+          .toPromise()
+          .then((result) => {
+            if (!result.stale /* XXX can this happen? */) {
+              set(operationResultAtom, result)
+            }
+          })
```

Comment:
@kitten I'm not sure if I understand this correctly. So, in imperative updates, can we use `toPromise()`? Or should we subscribe? (but then how can we unsubscribe)
Also, this should work both pause=true and pause=false, right?

Reply:
Well, not really; They're started imperatively but what you're changing is the source, so the most sure-fire way to make this work is to make the `source` (created using `client.executeQuery` or `client.query`) swappable itself, so that if it's set it swaps out the subscription and resubscribes to the new source. That'd make it very similar to how `@urql/vue` works. However, as discussed before, are we sure that we need to bring back these imperative triggers? In theory people could also swap out the entire atom themselves 🤔

Reply:
> However, as discussed before, are we sure that we need to bring back these imperative triggers? In theory people could also swap out the entire atom themselves

I thought `pause` is for imperative triggers. Do you mean we can drop supporting `pause`?

Reply:
It's a possibility! `pause` and the `executeQuery()` function on bindings (in the React/Preact/Vue bindings) exist because it's impossible to pause hooks once they're set up, however with Jotai's atoms that's not necessarily true unless they're only set up once. So it depends, but if a user, say, is encouraged to swap out the atom instead then there's no need for `pause`.